ds701_book/20-Recommender-Systems.qmd

```diff
@@ -404,169 +566,232 @@ The strategy for solving is:
 
 Each of the three steps will reduce the overall error.   So we iterate over them until convergence.
 
+---
+
 The last issue is that the matrix $R$ is not dense - in reality we only have a small subset of its entries.
 
 We simply need to adapt the least-squares solution to only consider the entries in $R$ that we know.
 
+---
+
 As a result, the actual calculation is:
 
 Step 1:
 
-$$ \mu = \frac{\sum_{(u, i) \in R} (r_{ui} - \alpha_u - \beta_i)}{|R|} $$
+$$
+\mu = \frac{\sum_{(u, i) \in R} (r_{ui} - \alpha_u - \beta_i)}{|R|} 
+$$
 
 Step 2: 
 
-$$ \alpha_u = \frac{\sum_{i \in R(u)}(r_{ui} - \mu - \beta_i)}{\lambda + |R(u)|} $$
+$$ 
+\alpha_u = \frac{\sum_{i \in R(u)}(r_{ui} - \mu - \beta_i)}{\lambda + |R(u)|} 
+$$
 
 Step 3:
     
-$$ \beta_i = \frac{\sum_{u \in R(i)}(r_{ui} - \mu - \alpha_u)}{\lambda + |R(i)|} $$
+$$ 
+\beta_i = \frac{\sum_{u \in R(i)}(r_{ui} - \mu - \alpha_u)}{\lambda + |R(i)|} 
+$$
 
 Step 4: If not converged, go to Step 1.
 
+---
+
 Now that we have the biases learned, we can do a better job of estimating correlation:
 
-$$ \hat{\rho}_{ij} = \frac{\sum_{u\in U(i,j)}(r_{ui} - b_{ui})(r_{uj}-b_{uj})} 
-{\sqrt{\sum_{u\in U(i,j)}(r_{ui} - b_{ui})^2\sum_{u\in U(i,j)}(r_{uj}-b_{uj})^2}} $$
+$$ 
+\hat{\rho}_{ij} = \frac{\sum_{u\in U(i,j)}(r_{ui} - b_{ui})(r_{uj}-b_{uj})} 
+{\sqrt{\sum_{u\in U(i,j)}(r_{ui} - b_{ui})^2\sum_{u\in U(i,j)}(r_{uj}-b_{uj})^2}} 
+$$
 
 where 
+
 * $b_{ui} = \mu + \alpha_u + \beta_i$, and
 * $U(i,j)$ are the users who have rated both $i$ and $j$.
 
+---
+
 And using biases we can also do a better job of estimating ratings:
 
-$$ \hat{r}_{ui} = b_{ui} + \frac{\sum_{j \in n_k(i, u)} s_{ij}(r_{uj} - b_{uj})}{\sum_{j \in n_k(i, u)} s_{ij}} $$
+$$ 
+\hat{r}_{ui} = b_{ui} + \frac{\sum_{j \in n_k(i, u)} s_{ij}(r_{uj} - b_{uj})}{\sum_{j \in n_k(i, u)} s_{ij}} 
+$$
 
 where 
+
 * $b_{ui} = \mu + \alpha_u + \beta_i$, and
 * $n_k(i, u)$ are the $k$ nearest neighbors to $i$ that were rated by user $u$.
 
-### Assessing CF
+## Assessing CF
 
 This completes the high level view of CF.
 
 Working with user-user similarities is analogous.
 
 Strengths:
+
 * Essentially no training.
     * The reliance on $k$-nearest neighbors helps in this respect.
 * Easy to update with new users, items, and ratings
 * Can be explained to user: 
     * "We recommend _Minority Report_ because you liked _Blade Runner_ and _Total Recall._"
 
 Weaknesses:
+
 * Accuracy can be a problem
 * Scalability can be a problem (think $k$-NN)
```

Comment:
For the weaknesses:

can we be more specific about when accuracy is a problem, is it the algorithm that doesn't converge well? Or is it a poor recommendation?

for scalability, is it just that large problems are intractable?